### DISCUSSION TITLE
Renderer-Agnostic Layout Engine?

### DISCUSSION THREAD

**Discussion by pha3z** [QuestPDF/QuestPDF#78] [Ideas]
I'm perusing the source code.  Bit tough to get acclimated to an entire code base, but the code looks very well architected.

I am wondering.  What do you think is the viability of making the layout engine completely renderer-agnostic?  I know there are some places where there are references to Skia types.  If they are few, it seems like we could generalize them to 'object' type to make the layout tree renderer-agnostic.  For the code that does the drawing, it could cast objects (like SKCanvas) to the expected Skia type.  For consumers of the fluent builder, we could use render-specific extension methods to dress up the frontend a bit.

I know working with general objects like this is ugly. But in my own libraries where I support multiple providers, it was the only elegant way I could ensure robustness for all providers since its a nightmare to try to unite them under a single interface.

Thoughts?

**Comment by pha3z** [comment]
FWIW, there are several GUI libraries that generate textured triangle lists that you can then draw on any graphics device.  They're very nice.  I'm thinking it would be useful to get similar abstraction here.  Then the layout engine could be used to draw and render other file formats.

The difference between this and the GUI libraries is that we don't want to produce any primitives.  Instead we just produce a layout tree and let an implementor be responsible for drawing the commands to render the tree.

**Comment by pha3z** [comment]
So far it looks like a major lynchpin is that all the elements inherit from Element class, which contains a Draw() method and an ICanvas property. That makes the element tree tightly coupled to rendering.

**Comment by pha3z** [comment]
Hrmm... :\

Yeah I'm looking at BinaryStack.Draw() which calls Draw() on its children.   For this to be renderer agnostic, I expect the BinaryStack (and every other Element also) would probably need to have a GetChildren() method that returns a list of children.  Each child would need a GetLayoutDescriptor() method that returns absolute position & width for the child, which the renderer would combine with other child data to decide exactly what and how to draw a representation.

The tree and its elements would be responsible for positioning and sizing themselves. An external visitor would be able to recurse the tree and render the result.  Implementation details -- such as rendering of text -- would be implemented by the external visitor.

Also, to avoid dynamic problems, the tree layout would need to be set-in-stone before the renderer recurses it.  So maybe after tree construction, an UpdateLayout() method on the tree could be called that goes through all the elements and sets their absolute sizes and widths.   Now its basically just a set of arguments for drawing commands -- the implementations of which are responsibility of the external visitor.

**Comment by MarcinZiabek** [comment]
This is a really interesting question 😁 Let's assume that we want to follow and make the layout engine renderer-agnostic. In such a case, we can pick different layers where we want to introduce generalization. For example, the ICanvas interface abstracts away rendering logic in the vast majority of places. It produces a stream of low-level rendering commands that need to be executed in order to display the result. Another idea is to decouple the element tree structure from the layouting engine. For each element, we would need to create a connected renderer object. This was my idea at the very beginning but I decided to skip it to keep everything simpler.

All things considered, before even thinking about code design and architecture, we need to answer the question: why? This layouting engine is tightly coupled with the document generation world, especially with the paging functionality. At this point, I don't see any other places where it can be successfully used. Any propositions?

**Comment by pha3z** [comment]
> This layouting engine is tightly coupled with the document generation world, especially with the paging functionality. 

It would _not_ make sense to use it for a UI system, because of the responsive nature of UIs.  Your layout system appears to work around fixed-size elements -- which makes some sense in a physical media document.  So it applies to all physical print media well.

I think my interest in decoupling it is primarily due to the fact that its backed by Skia, which completely fails us in the CMYK color space.  Given that this library is perfectly geared at physical documents which people want to use for print media, it would seriously ratchet up the awesomeness and professional value if it allowed generation of CMYK media. The obstacle in our way is the graphics device library... the renderer.

> Another idea is to decouple the element tree structure from the layouting engine. For each element, we would need to create a connected renderer object. This was my idea at the very beginning but I decided to skip it to keep everything simpler.

Yes, this is the approach I proposed in my former comment.  Honestly, I tend to favor anemic patterns as they are more robust to design changes (like the one we are now discussing).  If the tree was constructed as a data tree and wasn't responsible for rendering itself, it opens a lot of doors.

**Comment by MarcinZiabek** [comment]
> It would not make sense to use it for a UI system, because of the responsive nature of UIs. Your layout system appears to work around fixed-size elements -- which makes some sense in a physical media document. So it applies to all physical print media well.
> 
> I think my interest in decoupling it is primarily due to the fact that its backed by Skia, which completely fails us in the CMYK color space. Given that this library is perfectly geared at physical documents which people want to use for print media, it would seriously ratchet up the awesomeness and professional value if it allowed generation of CMYK media. The obstacle in our way is the graphics device library... the renderer.

Are there any alternatives that can help with generating PDFs and also support CMYK color space? I am mad enough (yet) to generate PDFs from scratch 🙉

> Yes, this is the approach I proposed in my former comment. Honestly, I tend to favor anemic patterns as they are more robust to design changes (like the one we are now discussing). If the tree was constructed as a data tree and wasn't responsible for rendering itself, it opens a lot of doors.

I am still not sure if the element tree alone helps much here. It may slightly help with generating Word documents in OpenXML. But still... without the layouting algorithm, the element tree itself is still not much. 

I am sorry for my temporary lack of optimism. I spent the last two weeks fighting with the Telerik library for Word (docx) document creation and I lost. It has just too many bugs and is not stable at all. I think the reason is that authors tried to recreate the Word structure themselves and were hit by the complexity. Finally, I switched to OpenXML with great success. Still, two weeks are lost 😭

**Comment by pha3z** [comment]
> I am still not sure if the element tree alone helps much here. It may slightly help with generating Word documents in OpenXML. But still... without the layouting algorithm, the element tree itself is still not much.

I thought that the layout algorithm is built within the tree.  I guess I haven't examined enough of the code yet.

Regardless, it doesn't really matter the details of the data.  Layout does not have to be tied to rendering.  If the layout algorithm is a visitor, you could produce a completely separate structure with the absolute rendering commands and then hand those off to the renderer.

Or if the layout algorithm works directly on the tree or within the tree, it could mutate the tree by setting all the absolute positions and widths etc on the tree.  Then the renderer can visit that.

**Comment by MarcinZiabek** [comment]
It depends on how you think about the code. Right now the element tree and layouting logic are represented by the same set of classes. I decided to go this path, as it makes development easier and code slightly cleaner. However, it is possible to split both apart, have an element tree with pure data, then a layouting layer and then a rendering layer.

At some point, I may decide to split them. I would like to extract away the state from the element and keep it in the monad. This will allow for more agile layouts, e.g. text columns (like in newspaper).

**Comment by pha3z** [comment]
> Are there any alternatives that can help with generating PDFs and also support CMYK color space? I am mad enough (yet) to generate PDFs from scratch

I have spent a ridiculous amount of hours searching for answers to that.   Almost everything I've found is commercial software.  And even worse, nearly all of it I've seen bills itself as a magic "Convert X to PDF" or "Convert PDF to X".  I haven't found anything that clearly says "Here, this is a library that will give you some data structures for knifing through a tokenized PDF"  or anything like that.  I think the PDF spec is just too complicated and overwhelming and nobody is tackling it except for a handful of hardcore folks that are want to get way into the weeds.

I actually looked at Microsoft's XPS format -- omg wow so much better!!!  Its sad that XPS has not become widely adopted.

Some pdf oriented tools:
- cairo
- poppler

But almost everything that I see that is open source falls back on Ghostscript.

**Reply by MarcinZiabek** [reply]
Over the last two years, I developed a strong aversion towards any products that perform a magic conversion between millions of formats. It usually does not work very well, if at all. But hey, this is a strong selling point - buy once and use for everything.

If anything, creating your own ICanvas object, is the most reasonable option, something worth trying. It has a handful of methods, one method per rendering command. It is not very complicated.

**Comment by pha3z** [comment]
From everything I've gleaned, the fundamental albatross with PDF is that it utilizes Postscript internally for some stuff.   I had no idea what Postscript even was before I got into this project.  I came from the 3D Rendering world where we think of everything as just a bunch of geometry lists that we do mutations on.  When I found out the historical way of storing vector data has been a turing-complete language, I about fell over.  It's just absurd.  SVG fixed that, but SVG didn't bother to go forward with their CMYK extension.  So we're in this terrible place now.

So you thought I'd go off the rails if I tried to parse Postscript from an EPS.  I did it.  It took me about 4 hours, using a tokenization library I wrote earlier in the year.  I took all the vector data in an EPS, translated it, and dumped out a new EPS.   It's not that bad because the programs that generate EPS aren't trying to generate "programs".  All they're doing is generating basic commands like "curveto" and "lineto" and "eofill" to get their data into a packaged format.  Obviously, its in the developers' best interests to keep their own code as simple as possible so they use the minimal options to generate a complete EPS file.

**Comment by pha3z** [comment]
> If anything, creating your own ICanvas object, is the most reasonable option, something worth trying. It has a handful of methods, one method per rendering command. It is not very complicated.

Ok I'm looking at your code now to get a feel the full scope of what that implies.

**Comment by MarcinZiabek** [comment]
I am thinking about [this one](https://github.com/QuestPDF/QuestPDF/blob/main/QuestPDF/Drawing/SkiaCanvasBase.cs).

**Comment by pha3z** [comment]
I don't think it'll work for my purposes. I really need a way to have arbitrary data (like an EPS document or a bitmap file) as the object for an element in the tree.  I specifically want to be able to stick a vector object in a container and then have my renderer be able to render that vector using the layout constraints dictated by the container.

If ICanvas had a "RenderElement(Element)" method, I'd be able to do.  As it is, I don't see how.  It's just too literal as an operator for drawing commands.

**Comment by pha3z** [comment]
So if this is the only way to do it.. I guess this is how it'd work:

I'd have to implement ICanvas and I'd have to add a method that allows rendering of the given element data.  Then I'd need to create a new Element that stores the specific kind of data needed to render, and that Element would need to call the special method on ICanvas --- or cast ICanvas to a specific implementation that has the known method.

Feelin' the pain of inheritance from a rich model here. XD

**Comment by pha3z** [comment]
"internal abstract class Element : IElement"

It's internal!  Can a library consumer even inherit from an internal class?

**Comment by MarcinZiabek** [comment]
Well... taking into account that the PDF format is also a set of commands, e.g. drawing text, lines, shapes, the set of methods inside the ICanvas interface is well suited for your purpose. 

> It's internal! Can a library consumer even inherit from an internal class?

Yes, many things are internal, therefore not available to consumers. I suggest downloading a source and just trying to alter it 😀

**Comment by pha3z** [comment]
I'm going to mod the lib for my own purposes.

I think its too much to ask to make Element public.   Since that was never your design intention, you'd be opening the door to a lot of breaking changes if people took dependencies on the internals of how your library works.

And I realize its probably just way too tall of an order to decouple rendering from layout at the present time.   I'm crazy enough, I might actually attempt it myself, but that's an enormous change to the lib.  May constitute more of a Version2 fork style development.

**Comment by MarcinZiabek** [comment]
I am still not sure why you would like to perform this change and why ICanvas is not enough. Would you like to describe your plans more precisely?

**Comment by pha3z** [comment]
For instance: you have an Image element already.
What I really need is an EpsElement or a VectorElement or even just a generic DataElement to which I can assign an Eps file.  When that element is rendered, the result is going to be the source vector art transformed to fit into the container defined by the Element.  To do that, the Canvas needs to understand the data (in this case an Eps file) in the Element so it can directly transform it and write it into the output.

So I need a parallel method to DrawImage(SKImage).  My method would be DrawEps(Eps)  or DrawObject(object).  My particular element would have to invoke the new method.

**Comment by pha3z** [comment]
I have a few ideas to enable rendering custom data types using custom canvas implementation.  Both ideas require creating a custom ObjectElement class.  The first variation requires an Action to handle drawing. The second variation requires modifying the ICanvas interface.

**Idea 1**

```
public class ObjectElement : Element
{
     public object Object;
     public Action<ICanvas, ObjectElement, Size> OnDraw;
}
```

When ObjectElement.Draw() is invoked, it invokes OnDraw and passes the ICanvas, 'this', and availableSize.  Consumer's Action delegate would:

- Cast ICanvas to specific implementation
- Cast ObjectElement.Object to known type and render it with user-defined logic.

**Idea 2**

```
public class ObjectElement : Element
{
     public object Object;
}
```

Add this method to ICanvas:
ICanvas.DrawObject(object object, Position position, Size size);

Implementor would be responsible custom implementation of DrawObject() and invoking object.GetType() and mapping to the custom drawing code.

**Comment by pha3z** [comment]
I really like the delegate version because it would allow user to conveniently work with SkCanvas directly (not requiring custom canvas implementation). User's imagination would be the only limit.

**Comment by pha3z** [comment]
~~Oh! I'm posting too fast before thinking through things. In the variation that adds a method to ICanvas, you could add a method to the top-level API that lets user specify a delegate for handling ObjectElement draw calls.  This keeps delegate references out of the fluent builder.  Much cleaner and would probably work better in most situations.~~

**Comment by MarcinZiabek** [comment]
Don't you think about [the Canvas element](https://www.questpdf.com/documentation/api-reference.html#canvas)? This is the only place where QuestPDF exposes SkiaSharp in the public API. I have done this to give full control over the document's content. If the developer needs something special, he just gets access to the low-level drawing API. This way, we can[ implement vector-based charts](https://www.questpdf.com/documentation/patterns-and-practices.html#implementing-charts) with other SkiaSharp-powered libraries.

**Comment by pha3z** [comment]
> Don't you think about the Canvas element? 

Ok, I guess that works.  I had to think about it for a bit.  So when declaring the canvas, I would inject a delegate that captures the instance of data I need to work with and performs the drawing of the captured instance.

I guess that's pretty much what I was describing in Idea 1.

It's slightly inconvenient and ugly in that there's an extra level of indirection in the fluent chain.... because we're providing a full delegate every time we declare a Canvas element.  This is opposite to Idea 2 where we only provide the data and allow a predefined master drawing delegate to figure out what the data is and how to draw it.

Canvas itself has a problem for renderer-agnostic use, though.  Internally it casts to Drawing.SkiaCanvasBase and when it invokes DrawOnCanvas it passes an SKCanvas.
[https://github.com/QuestPDF/QuestPDF/blob/main/QuestPDF/Elements/Canvas.cs](https://github.com/QuestPDF/QuestPDF/blob/main/QuestPDF/Elements/Canvas.cs)

**Comment by pha3z** [comment]
Also... ICanvas is internal.
"internal interface ICanvas {}"

This seems like the type of thing that would be public for an implementor.  Is there a reason why it needs to remain internal?

**Comment by pha3z** [comment]
I added Pull Request #80 to demonstrate my idea.  However I have an outstanding issue:

I don't understand how your positioning system works yet.  Right now the call to Canvas.DrawObject() passes Position.Zero.  But I don't think that's correct.  How do you pass the correct position to the ICanvas so it knows where to draw the element?

**Comment by pha3z** [comment]
If I understand what's going on, it looks like you're setting the Translate values in the SKMatrix in order for the canvas to know where to draw.

Either I'm not understanding this system correctly, or there's just too much tight coupling to the Skia rendering system to make it practical to do a drop-in a replacement ICanvas. I think I'm going to create a new library that is an implementation of only the layout engine without any rendering.

**Comment by MarcinZiabek** [comment]
That's correct. I am using skia matrix operations to position the origin in the desired place, instead of calculating positions myself. This makes the code much faster (similar to game engines). Any rendering ICanvas replacement can perform similar operations 😁

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
